pkg/machines/v1alpha1/machineset_types.go

```diff
@@ -0,0 +1,167 @@
+	// +optional
+	Template MachineTemplateSpec `json:"template,omitempty"`
+}
+
+// MachineTemplateSpec describes the data a machine should have when created from a template
+type MachineTemplateSpec struct {
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired behavior of the machine.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec MachineSpec `json:"spec,omitempty"`
+}
+
+// MachineSetStatus defines the observed state of MachineSet
+type MachineSetStatus struct {
+	// Replicas is the most recently observed number of replicas.
+	Replicas int32 `json:"replicas"`
+
+	// The number of replicas that have labels matching the labels of the machine template of the MachineSet.
+	// +optional
+	FullyLabeledReplicas int32 `json:"fullyLabeledReplicas,omitempty"`
+
+	// The number of ready replicas for this MachineSet. A machine is considered ready when the node has been created and is "Ready".
+	// +optional
+	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+	// The number of available replicas (ready for at least minReadySeconds) for this MachineSet.
+	// +optional
+	AvailableReplicas int32 `json:"availableReplicas,omitempty"`
+
+	// ObservedGeneration reflects the generation of the most recently observed MachineSet.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// In the event that there is a terminal problem reconciling the
+	// replicas, both ErrorReason and ErrorMessage will be set. ErrorReason
+	// will be populated with a succinct value suitable for machine
+	// interpretation, while ErrorMessage will contain a more verbose
+	// string suitable for logging and human consumption.
+	//
+	// These fields should not be set for transitive errors that a
+	// controller faces that are expected to be fixed automatically over
+	// time (like service outages), but instead indicate that something is
+	// fundamentally wrong with the MachineTemplates's spec or the configuration of
+	// the machine controller, and that manual intervention is required. Examples
+	// of terminal errors would be invalid combinations of settings in the
+	// spec, values that are unsupported by the machine controller, or the
+	// responsible machine controller itself being critically misconfigured.
+	//
+	// Any transient errors that occur during the reconcilation of Machines
+	// can be added as events to the MachineSet object and/or logged in the
+	// controller's output.
+	// +optional
+	ErrorReason *common.MachineSetStatusError `json:"errorReason,omitempty"`
+	// +optional
+	ErrorMessage *string `json:"errorMessage,omitempty"`
+}
+
+// Validate checks that an instance of MachineSet is well formed
+//func (MachineSetStrategy) Validate(ctx request.Context, obj runtime.Object) field.ErrorList {
```

Comment:
I commented this out because `MachineSetStrategy` transiently depends on apiserver builder

Reply:
We could completely remove it?
With the use of CRD's, this code does not get used anyway?

Reply:
For the validation, yes